### PR TITLE
Check encrypted files with SHA-512/256 checksums

### PR DESCRIPTION
**Pull request by mzuther**
### What does this PR do?

This pull request adds three contributed hooks (`post_encrypt`, `post_list` and `post_status`) which create, check and list encrypted files using SHA-256 checksums. The PR adds a dependency for `shasum`, `sha256sum` or `gsha256sum`.

There is also a small change in the `yadm` script.  To avoid code duplication, `invoke_hook` exports `YADM_DIR`, the array `ENCRYPT_INCLUDE_FILES` to `YADM_ENCRYPT_INCLUDE_FILES` and some helper functions.

Exporting the array *removes support for file names containing newline characters*. Which is probably fine – using (as well as supporting) file names with newlines opens a can of worms.

### What issues does this PR fix or reference?

#126 

### Previous Behavior

There is no way of checking whether encrypted files have changed.

### New Behavior

The `yadm encrypt` hook creates a file `files.checksums` containing checksums next to `files.gpg`. If encrypted files change, the `yadm status` displays the files with mismatched checksums. `yadm list [-a]` lists encrypted files.

### Have [tests][1] been written for this change?

No

I could't think of a way to test this.  If you have an idea, I'll write some tests.

### Have these commits been [signed with GnuPG][2]?

Yes

[1]: https://github.com/TheLocehiliosan/yadm/blob/master/.github/CONTRIBUTING.md#test-conventions
[2]: https://help.github.com/en/articles/signing-commits
[3]: https://github.com/TheLocehiliosan/yadm/blob/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: contrib/hooks/post_encrypt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        echo
+        echo "WARNING: \"shasum\" does not support SHA-512/256.  No checksums were created."
+        exit $YADM_HOOK_EXIT
+    fi
+
+    # empty checksum file
+    echo -n > $YADM_CHECKSUMS
+
+    # calculate checksums for encrypted files
+    for included in ${YADM_ENCRYPT_INCLUDE_FILES[@]}; do
```
Comment (rasa) [inline contrib/hooks/post_encrypt]: To support spaces in filenames: `”${YADM_ENCRYPT_INCLUDE_FILES[@]}”`

**Comment by TheLocehiliosan** [comment]
Thanks for the contribution. I can easily create some tests for the actual yadm changes. I'm going to suggest one other thing. Can you also have `YADM_DIR` exported for hooks, and then have the hooks use that variable instead of `dirname`? I've found that `dirname` is not particularly portable. It would make the hooks work for more people.

File: contrib/hooks/post_encrypt (hunk 1 above)
Comment (mzuther) [reply contrib/hooks/post_encrypt]: Thanks, fixed. Quotes drive me nuts in Bash. :)

**Comment by rasa** [comment]
`shasum` is less commonly found than `sha256sum`. See https://unix.stackexchange.com/a/426838. You might consider trying it first and then falling back to using shasum. Perhaps, something like:
```
if ! command -v sha256sum &>/dev/null; then
   function sha256sum() { shasum -a 256 "$@" ; }
fi
```

**Comment by mzuther** [comment]
> `shasum` is less commonly found than `sha256sum`. See https://unix.stackexchange.com/a/426838.

Thanks for the suggestion!

However, I think that it is the other way round – please look at the [link](https://github.com/ESGF/esg-search/issues/84) in the link you provided. I have also found a man page for [shasum](https://ss64.com/osx/shasum.html). Am I correct?

But your comment wasn't in vain. After reading the above man page, I have changed the algorithm to SHA-512 as the newer ones don't seem to be supported on OS X.

**Comment by rasa** [comment]
> However, I think that it is the other way round – please look at the [link](https://github.com/ESGF/esg-search/issues/84) in the link you provided. I have also found a man page for [shasum](https://ss64.com/osx/shasum.html). Am I correct?

On MacOS you are. On Linux and it’s variants (Cygwin, Msys, git-for-windows, etc.), sha256sum is more common. I would consider cycling thru different options, such as:

`sha256sum`
`gsha256sum`
`shasum -a 256`
`gpg --print-md sha256 "$@" | cut -d: -f 2- | tr -d ' '`
`openssl sha256 "$@" | awk '{print $2}'; }` 

You might also consider searching in `/usr/local/bin` as well as the PATH, as this is a common location for `gsha256sum`

**Comment by mzuther** [comment]
I added checks for `shasum`, `sha256sum` and `gsha256sum` (in that order). `gpg` and `openssl` are nice options, but they complicate things.

Also, I didn't add extra checks for `/usr/local/bin`. While this is a nice idea, I think this is going a bit far. I feel that users have some responsibilities, and correctly setting the PATH is one of them.

Note that my scripts break if file names contain newline characters. This keeps the scripts simple, and not having newlines in file names is also one of the responsibilities of users.

**Comment by mzuther** [comment]
@TheLocehiliosan I'll handle the `dirname` issue tomorrow.

It might also be nice to extend `yadm list -a`. Let's see... :smile:

**Comment by mzuther** [comment]
Done. :smile:

Comments or ideas?

File: contrib/hooks/post_encrypt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+if [ $ERROR_CODE -ne 0 ]; then
+    # return original exit status of yadm command
+    exit "$YADM_HOOK_EXIT"
+fi
+
+# empty (or create) checksum file
+echo -n > "$YADM_CHECKSUMS"
+
+# calculate checksums for encrypted files
+for included in ${YADM_ENCRYPT_INCLUDE_FILES[*]}; do
```
Comment (rasa) [inline contrib/hooks/post_encrypt]: Shouldn’t this be `${YADM_ENCRYPT_INCLUDE_FILES[@]}`?

File: contrib/hooks/post_list [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        REL_PATH=$(relative_path "$PWD" "$YADM_HOOK_WORK/$filename")
+
+        if [ "$REL_PATH" = "${REL_PATH##../}" ]; then
+            echo "$REL_PATH"
+        fi
+    # list all files
+    else
+        echo "$filename"
+    fi
+done <<< "$(cat "$YADM_CHECKSUMS")"
```
Comment (rasa) [inline contrib/hooks/post_list]: This can be simplified to `<"$YADM_CHECKSUMS"`

File: contrib/hooks/post_status [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+if [ ! -f "$YADM_CHECKSUMS" ]; then
+    exit "$YADM_HOOK_EXIT"
+fi
+
+# get checksum command
+CHECKSUM_COMMAND=$(get_checksum_command)
+ERROR_CODE=$?
+
+# no command found
+if [ $ERROR_CODE -ne 0 ]; then
```
Comment (rasa) [inline contrib/hooks/post_status]:
```
ERROR_CODE=$?

 # no command found
 if [ $ERROR_CODE -ne 0 ]; then
```
Can be simplified to:
```
 # no command found
 if (($?)); then
```

File: yadm [hunk 5]
```diff
@@ -1819,6 +1833,13 @@ function auto_bootstrap() {
 
 }
 
+# ****** Helper Functions ******
+
+function join_string {
+    local IFS="$1"
+    echo "${*:2}"
```
Comment (rasa) [inline yadm]: This may fail on files that start with a `-`. Consider using `printf`  instead. See https://askubuntu.com/a/467756

Comment (rasa) [reply yadm]: Or do you intend to call the `echo` function defined below?

File: contrib/hooks/post_encrypt (hunk 2 above)
Comment (mzuther) [reply contrib/hooks/post_encrypt]: You're right, thanks!

File: contrib/hooks/post_list (hunk 3 above)
Comment (mzuther) [reply contrib/hooks/post_list]: Oops, fixed.

File: contrib/hooks/post_status (hunk 4 above)
Comment (mzuther) [reply contrib/hooks/post_status]: Neat, I didn't know this one.  Changed.

File: yadm (hunk 5 above)
Comment (mzuther) [reply yadm]: I'm really learning something here! :smile: Thanks!!!

I got rid of all other `echo` instances, too, and changed them to `printf`. Just to get used to it...

**Comment by TheLocehiliosan** [comment]
@mzuther Thanks for the contribution. I've added some organization to the contrib/hooks directory, and added tests for the changes to yadm itself.

This has been merged into the develop branch.

**Comment by mzuther** [comment]
You're more than welcome!

**Comment by rasa** [comment]
> I added checks for shasum, sha256sum and gsha256sum (in that order). gpg and openssl are nice options, but they complicate things.

I would suggest that gpg is already a dependency, so that could be the default, and then fallback on searching for an alternative, if gpg is not found for some reason.

**Comment by mzuther** [comment]
@rasa I do appreciate your comment and also understand the reasoning behind it.

In this case, however, I think that adding a (non-essential) dependency is the better way. `shasum` and its relatives have only one job to do and use a (semi-)standardized hash file that is unlikely to change. `gpg` has a different job and its output might change over time. Thus, removing a dependency now ('shasum') could well turn into a liability later (for example a requirement for checking `gpg` versions).

**Pull request merged**